File: src/main/kotlin/ComputeAverageIncome.kt
import EmployeeServices.Companion.lookupAllEmployees
import kotlin.random.Random.Default.nextInt
import kotlin.system.measureTimeMillis

data class Address(
        val street: String,
        val postalCode: String,
        val city: String,
        val country: String)

data class Employee(
        val firstName: String,
        val lastName: String,
        val address:
        Address,
        val salary: Long)

class EmployeeServices {

    companion object {
        private val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')

        fun lookupAllEmployees(numberOfAllEmployees: Long): Sequence<Employee> {
            return (1L..numberOfAllEmployees)
                    .asSequence()
                    .map { createRandomEmployee() }
        }

        private fun createRandomEmployee(): Employee =
                Employee(
                        createRandomStringOf80Chars(),
                        createRandomStringOf80Chars(),
                        Address(
                                createRandomStringOf80Chars(),
                                createRandomStringOf80Chars(),
                                createRandomStringOf80Chars(),
                                createRandomStringOf80Chars()
                        ), 1000L
                )

        private fun createRandomStringOf80Chars() =
                (1..80)
                        .map { nextInt(0, charPool.size) }
                        .map(charPool::get)
                        .joinToString("")


        fun computeAverageIncome(employees: Sequence<Employee>): Double {
            val (nrOfEmployees, sumOfSalaries) = employees
                    .fold(Pair(0L, 0L),
                            { (counter, sum), employee ->
                                Pair(counter + 1, sum + employee.salary)
                            })
            return sumOfSalaries.toDouble() / nrOfEmployees.toDouble()
        }
    }
}

fun main() {
    val nrsOfEmployees = listOf(1000L, 10000L, 100000L, 1000000L)
    nrsOfEmployees.forEach { nr: Long ->
        val timeNeeded = measureTimeMillis {
            EmployeeServices.computeAverageIncome(lookupAllEmployees(nr))
            Runtime.getRuntime().gc()
        }
        println("timeNeeded = $timeNeeded ms")
    }
}

File: src/main/Haskell/compute-average-income/src/ComputeAverageIncome.hs
module ComputeAverageIncome
    ( computeAverageIncomeOfAllEmployees
    , createRandomStringOf80Chars
    , createRandomAddress
    , createRandomEmployee
    , lookupAllEmployees
    , computeAverageIncome
    , Address( Address )
    , Employee( Employee)
    ) where

import System.Random
import Control.Monad

data Address = Address
  { street:: String
  , postalCode:: String
  , city:: String
  , country:: String
  } deriving (Eq, Show)

data Employee = Employee
  { firstName:: String
  , lastName:: String
  , address:: Address
  , salary:: Int
  } deriving (Eq, Show)

chars = ['a'..'z'] ++ ['A'..'Z'] ++ ['0'..'9']
nrOfChars = length chars

createRandomStringOf80Chars :: IO String
createRandomStringOf80Chars = do
  random <- newStdGen
  return $
    take 80
      $ map (\i -> chars !! i)
      $ randomRs ( 0,nrOfChars -1) random

createRandomAddress :: IO Address
createRandomAddress = do
  streetV <- createRandomStringOf80Chars
  postalCodeV <- createRandomStringOf80Chars
  cityV <- createRandomStringOf80Chars
  countryV <- createRandomStringOf80Chars
  return Address
    { street = streetV
    , postalCode = postalCodeV
    , city = cityV
    , country = countryV
    }

createRandomEmployee :: IO Employee
createRandomEmployee = do
  firstNameV <- createRandomStringOf80Chars
  lastNameV <- createRandomStringOf80Chars
  addressV <- createRandomAddress
  let salaryV = 1000
  return Employee
    { firstName = firstNameV
    , lastName = lastNameV
    , address = addressV
    , salary = salaryV
}

lookupAllEmployees :: Int -> IO [Employee]
lookupAllEmployees numberOfAllEmployees =
  replicateM numberOfAllEmployees createRandomEmployee

computeAverageIncome :: [Employee] -> Float
computeAverageIncome employees =
  let (nrEmployees, sumOfAllSalaries) =
        foldr (\ employee (counter, sum)  -> (counter + 1, sum + (salary employee)))
              (0,0) employees
  in
    (fromIntegral sumOfAllSalaries) / (fromIntegral nrEmployees)

computeAverageIncomeOfAllEmployees :: Int -> IO (Float)
computeAverageIncomeOfAllEmployees numberOfAllEmployees = do
  employees <- lookupAllEmployees numberOfAllEmployees
  return $ computeAverageIncome employees